KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,407 @@
+"""
+    typeon_obj = get_typeon_obj(seq::Sequence, type::String)
+
+Returns the vector `typeon_obj` of objects (RF, Grad or ADC) where the object is on in a
+given sequence `seq` according to the `type` ∈ ["rf", "gr", "adc"] argument.
+"""
+function get_typeon_obj(seq::Sequence, type::String)
+    type == "rf"  && return [s.RF[1] for s ∈ seq if is_RF_on(s)]
+    type == "gr"  && return [g for s ∈ seq for g ∈ s.GR[:,1] if sum(abs.(g.A)) > 0]
+    type == "adc" && return [s.ADC[1] for s ∈ seq if is_ADC_on(s)]
+    return []
+end
+
+"""
+    typeunique_obj_id = get_typeunique_obj_id(typeon_obj::Vector)
+
+Returns the vector `typeunique_obj_id` of objects (RF, Grad or ADC) which are unique given a
+input vector `typeon_obj` of objects which are on.
+"""
+function get_typeunique_obj_id(typeon_obj::Vector)
+    typeunique_obj_id = []
+    id_cnt = 1
+    for obj ∈ typeon_obj
+        if all([!(obj ≈ obj_unique) for (obj_unique, _) ∈ typeunique_obj_id])
+            push!(typeunique_obj_id, [obj, id_cnt])
+            id_cnt += 1
+        end
+    end
+    return typeunique_obj_id
+end
+
+"""
+    obj = get_obj(s::Sequence, type::String)
+
+Returns the object (RF, Grad or ADC) of a Sequence `s` (ideally a one-block sequence)
+according to the input `type` ∈ ["rf", "gx", "gy", "gz", "adc"]).
+"""
+function get_obj(s::Sequence, type::String)
+    type == "rf"  && return s.RF[1]
+    type == "gx"  && return s.GR[1,1]
+    type == "gy"  && return s.GR[2,1]
+    type == "gz"  && return s.GR[3,1]
+    type == "adc" && return s.ADC[1]
+    return Sequence()
+end
+
+"""
+    type_blk_obj_id = get_type_blk_obj_id(seq::Sequence, type::String, unique_obj_id::Vector)
+
+Returns the vector `type_blk_obj_id` of vectors [blk, obj, id] for all the blocks `blk` of a
+sequence `seq` according to the `type` ∈ ["rf", "gx", "gy", "gz", "adc"] chosen. It is
+neccessary to add the input vector `unique_obj_id` which contains the uniques objects (RF,
+Grad, or ADC) with its repective ID.
+"""
+function get_type_blk_obj_id(seq::Sequence, type::String, unique_obj_id::Vector)
+
+    type_blk_obj_id = [[blk, get_obj(s, type), 0] for (blk, s) ∈ enumerate(seq)]
+    for boi ∈ type_blk_obj_id
+        for (obj_unique, id_unique) ∈ unique_obj_id
+            if boi[2] ≈ obj_unique
+                boi[3] = id_unique
+            end
+        end
+    end
+    return type_blk_obj_id
+end
+
+"""
+    output = magsign(v)
+
+Returns the extreme value (positive or negative) which has the maximum absolute value.
+"""
+function magsign(v)
+    a, b = extrema(v)
+    return -a > b ? a : b
+end
+
+"""
+Given an angle in the range [-1, 1], it returns unwrapped values in the range [-0.5, 0.5].
+"""
+function unwrap_angle(ang)
+    (ang <= -0.5 + 1e-7) && return ang + 1.0
+    (0.5 + 1e-7 < ang) && return ang - 1.0
+    return ang
```

Comment:
what is this? I think i fixed the necessity of using tricks like this in the code I sent to @beorostica . Also I dont like the abuse of `bool && return something`

Reply:
Addressed in 0d4d5e0